File: lib/notification_service/java/src/main/java/org/aiflow/notification/client/EmbeddedNotificationClient.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.aiflow.notification.client;

import com.google.common.util.concurrent.ThreadFactoryBuilder;
import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import org.aiflow.notification.conf.Configuration;
import org.aiflow.notification.entity.EventMeta;
import org.aiflow.notification.entity.SenderEventCount;
import org.aiflow.notification.proto.NotificationServiceGrpc;
import org.aiflow.notification.proto.NotificationServiceOuterClass;
import org.apache.commons.lang3.StringUtils;
import org.apache.commons.lang3.tuple.ImmutablePair;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.*;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

import static org.aiflow.notification.conf.Configuration.CLIENT_ENABLE_IDEMPOTENCE_CONFIG_DEFAULT_VALUE;
import static org.aiflow.notification.conf.Configuration.CLIENT_ENABLE_IDEMPOTENCE_CONFIG_KEY;
import static org.aiflow.notification.conf.Configuration.CLIENT_ID_CONFIG_DEFAULT_VALUE;
import static org.aiflow.notification.conf.Configuration.CLIENT_ID_CONFIG_KEY;
import static org.aiflow.notification.conf.Configuration.CLIENT_INITIAL_SEQUENCE_NUMBER_CONFIG_DEFAULT_VALUE;
import static org.aiflow.notification.conf.Configuration.CLIENT_INITIAL_SEQUENCE_NUMBER_CONFIG_KEY;
import static org.aiflow.notification.conf.Configuration.HA_CLIENT_LIST_MEMBERS_INTERVAL_MS_CONFIG_DEFAULT_VALUE;
import static org.aiflow.notification.conf.Configuration.HA_CLIENT_LIST_MEMBERS_INTERVAL_MS_CONFIG_KEY;
import static org.aiflow.notification.conf.Configuration.HA_CLIENT_RETRY_INTERVAL_MS_CONFIG_DEFAULT_VALUE;
import static org.aiflow.notification.conf.Configuration.HA_CLIENT_RETRY_INTERVAL_MS_CONFIG_KEY;
import static org.aiflow.notification.conf.Configuration.HA_CLIENT_RETRY_TIMEOUT_MS_CONFIG_DEFAULT_VALUE;
import static org.aiflow.notification.conf.Configuration.HA_CLIENT_RETRY_TIMEOUT_MS_CONFIG_KEY;

public class EmbeddedNotificationClient extends NotificationClient {

    private static final Logger logger = LoggerFactory.getLogger(EmbeddedNotificationClient.class);
    public static String ANY_CONDITION = "*";
    private static final String SERVER_URI = "localhost:50051";
    private final String namespace;
    private final Map<String, EventListener> threads;
    private final ExecutorService listMembersService;
    private ManagedChannel channel;
    private NotificationServiceGrpc.NotificationServiceBlockingStub notificationServiceStub;
    private Set<NotificationServiceOuterClass.MemberProto> livingMembers;
    private Boolean enableHa;
    private String currentUri;
    private String sender;
    private Long clientId;
    private AtomicInteger sequenceNum;
    private final Configuration conf;

    public EmbeddedNotificationClient(
            String target, String namespace, String sender, Properties properties)
            throws Exception {
        this.namespace = namespace;
        this.sender = sender;
        this.conf = new Configuration(properties);

        String[] serverUris = StringUtils.split(target, ",");
        if (serverUris.length > 1) {
            this.enableHa = true;
            boolean lastError = true;
            for (String serverUri : serverUris) {
                currentUri = serverUri;
                try {
                    initNotificationServiceStub();
                    lastError = false;
                    break;
                } catch (Exception e) {
                    continue;
                }
            }
            if (lastError) {
                logger.warn("Failed to initialize client");
            }
        } else {
            this.enableHa = false;
            this.currentUri = target;
            initNotificationServiceStub();
        }

        long clientId = this.conf.getLong(CLIENT_ID_CONFIG_KEY, CLIENT_ID_CONFIG_DEFAULT_VALUE);
        int initialSeqNum =
                this.conf.getInt(
                        CLIENT_INITIAL_SEQUENCE_NUMBER_CONFIG_KEY,
                        CLIENT_INITIAL_SEQUENCE_NUMBER_CONFIG_DEFAULT_VALUE);
        if (clientId < 0) {
            this.clientId = registerClient();
        } else {
            if (checkClientExists(clientId)) {
                this.clientId = clientId;
            } else {
                throw new Exception(
                        "Init notification client with a client id which have not registered.");
            }
        }
        this.sequenceNum = new AtomicInteger(initialSeqNum);

        threads = new HashMap<>();
        livingMembers = new HashSet<>();
        listMembersService =
                Executors.newSingleThreadExecutor(
                        new ThreadFactoryBuilder()
                                .setDaemon(true)
                                .setNameFormat("list-members-%d")
                                .build());
        listMembersService.submit(listMembers());
    }

    /**
     * *
     *
     * @return Id of registered client
     */
    private long registerClient() throws Exception {
        NotificationServiceOuterClass.ClientMeta.Builder builder =
                NotificationServiceOuterClass.ClientMeta.newBuilder();
        if (!StringUtils.isEmpty(namespace)) {
            builder.setNamespace(namespace);
        }
        if (!StringUtils.isEmpty(sender)) {
            builder.setSender(sender);
        }
        NotificationServiceOuterClass.RegisterClientRequest registerClientRequest =
                NotificationServiceOuterClass.RegisterClientRequest.newBuilder()
                        .setClientMeta(builder.build())
                        .build();
        NotificationServiceOuterClass.RegisterClientResponse registerClientResponse =
                notificationServiceStub.registerClient(registerClientRequest);
        if (registerClientResponse.getReturnCode()
                == NotificationServiceOuterClass.ReturnStatus.SUCCESS) {
            return registerClientResponse.getClientId();
        } else {
            throw new Exception(registerClientResponse.getReturnMsg());
        }
    }

    private boolean checkClientExists(long clientId) {
        NotificationServiceOuterClass.ClientIdRequest request =
                NotificationServiceOuterClass.ClientIdRequest.newBuilder()
                        .setClientId(clientId)
                        .build();
        NotificationServiceOuterClass.isClientExistsResponse response =
                notificationServiceStub.isClientExists(request);
        if (response.getReturnCode() == NotificationServiceOuterClass.ReturnStatus.SUCCESS
                && response.getIsExists()) {
            return true;
        } else {
            return false;
        }
    }

    protected static List<EventMeta> listAllEvents(
            NotificationServiceGrpc.NotificationServiceBlockingStub serviceStub,
            Long startTime,
            Long startOffset,
            Long endOffset,
            int timeoutSeconds)
            throws Exception {
        NotificationServiceOuterClass.ListAllEventsRequest.Builder builder =
                NotificationServiceOuterClass.ListAllEventsRequest.newBuilder();
        if (startTime != null) {
            builder.setStartTime(startTime.longValue());
        }
        if (startOffset != null) {
            builder.setStartOffset(startOffset.longValue());
        }
        if (endOffset != null) {
            builder.setEndOffset(endOffset.longValue());
        }

        NotificationServiceOuterClass.ListAllEventsRequest request =
                builder.setTimeoutSeconds(timeoutSeconds).build();
        NotificationServiceOuterClass.ListEventsResponse response =
                serviceStub.listAllEvents(request);
        return parseEventsFromResponse(response);
    }

    private static List<EventMeta> parseEventsFromResponse(
            NotificationServiceOuterClass.ListEventsResponse response) throws Exception {
        if (response.getReturnCode() == NotificationServiceOuterClass.ReturnStatus.SUCCESS) {
            List<EventMeta> eventMetas = new ArrayList<>();
            for (NotificationServiceOuterClass.EventProto eventProto : response.getEventsList()) {
                eventMetas.add(EventMeta.buildEventMeta(eventProto));
            }
            return eventMetas;
        } else {
            throw new Exception(response.getReturnMsg());
        }
    }

    private static ImmutablePair<Long, List<SenderEventCount>> parseEventCountFromResponse(
            NotificationServiceOuterClass.CountEventsResponse response) throws Exception {
        if (response.getReturnCode() == NotificationServiceOuterClass.ReturnStatus.SUCCESS) {
            List<SenderEventCount> senderEventCounts = new ArrayList<>();
            for (NotificationServiceOuterClass.SenderEventCountProto eventCountProto :
                    response.getSenderEventCountsList()) {
                senderEventCounts.add(SenderEventCount.buildSenderEventCount(eventCountProto));
            }
            return new ImmutablePair<>(response.getEventCount(), senderEventCounts);
        } else {
            throw new Exception(response.getReturnMsg());
        }
    }

    protected static NotificationServiceGrpc.NotificationServiceBlockingStub wrapBlockingStub(
            NotificationServiceGrpc.NotificationServiceBlockingStub stub,
            String target,
            Set<NotificationServiceOuterClass.MemberProto> livingMembers,
            Boolean haRunning,
            Integer retryIntervalMs,
            Integer retryTimeoutMs) {
        return NotificationServiceGrpc.newBlockingStub(
                        ManagedChannelBuilder.forTarget(target).usePlaintext().build())
                .withInterceptors(
                        new NotificationInterceptor(
                                stub,
                                target,
                                livingMembers,
                                haRunning,
                                retryIntervalMs,
                                retryTimeoutMs));
    }

    /** Select a valid server from server candidates as current server. */
    protected void selectValidServer() {
        boolean lastError = false;
        int listMemberIntervalMs =
                this.conf.getInt(
                        HA_CLIENT_LIST_MEMBERS_INTERVAL_MS_CONFIG_KEY,
                        HA_CLIENT_LIST_MEMBERS_INTERVAL_MS_CONFIG_DEFAULT_VALUE);
        for (NotificationServiceOuterClass.MemberProto livingMember : livingMembers) {
            try {
                currentUri = livingMember.getServerUri();
                initNotificationServiceStub();
                NotificationServiceOuterClass.ListMembersRequest request =
                        NotificationServiceOuterClass.ListMembersRequest.newBuilder()
                                .setTimeoutSeconds(listMemberIntervalMs / 1000)
                                .build();
                NotificationServiceOuterClass.ListMembersResponse response =
                        notificationServiceStub.listMembers(request);
                if (response.getReturnCode()
                        == NotificationServiceOuterClass.ReturnStatus.SUCCESS) {
                    livingMembers = new HashSet<>(response.getMembersList());
                    lastError = false;
                    break;
                } else {
                    lastError = true;
                }
            } catch (Exception e) {
                lastError = true;
            }
        }
        if (lastError) {
            logger.warn("No available server uri!");
        }
    }

    /** Initialize notification service stub. */
    protected void initNotificationServiceStub() {
        if (notificationServiceStub == null) {
            channel =
                    ManagedChannelBuilder.forTarget(
                                    StringUtils.isEmpty(currentUri) ? SERVER_URI : currentUri)
                            .usePlaintext()
                            .build();
            notificationServiceStub = NotificationServiceGrpc.newBlockingStub(channel);
        }
        if (enableHa) {
            int retryIntervalMs =
                    this.conf.getInt(
                            HA_CLIENT_RETRY_INTERVAL_MS_CONFIG_KEY,
                            HA_CLIENT_RETRY_INTERVAL_MS_CONFIG_DEFAULT_VALUE);
            int retryTimeoutMs =
                    this.conf.getInt(
                            HA_CLIENT_RETRY_TIMEOUT_MS_CONFIG_KEY,
                            HA_CLIENT_RETRY_TIMEOUT_MS_CONFIG_DEFAULT_VALUE);
            notificationServiceStub =
                    wrapBlockingStub(
                            notificationServiceStub,
                            StringUtils.isEmpty(currentUri) ? SERVER_URI : currentUri,
                            livingMembers,
                            enableHa,
                            retryIntervalMs,
                            retryTimeoutMs);
        }
    }

    /** List living members under high available mode. */
    protected Runnable listMembers() {
        return () -> {
            int listMemberIntervalMs =
                    this.conf.getInt(
                            HA_CLIENT_LIST_MEMBERS_INTERVAL_MS_CONFIG_KEY,
                            HA_CLIENT_LIST_MEMBERS_INTERVAL_MS_CONFIG_DEFAULT_VALUE);
            while (enableHa) {
                try {
                    if (Thread.currentThread().isInterrupted()) {
                        break;
                    }
                    NotificationServiceOuterClass.ListMembersRequest request =
                            NotificationServiceOuterClass.ListMembersRequest.newBuilder()
                                    .setTimeoutSeconds(listMemberIntervalMs / 1000)
                                    .build();
                    NotificationServiceOuterClass.ListMembersResponse response =
                            notificationServiceStub.listMembers(request);
                    if (response.getReturnCode()
                            == NotificationServiceOuterClass.ReturnStatus.SUCCESS) {
                        livingMembers = new HashSet<>(response.getMembersList());
                    } else {
                        logger.warn(response.getReturnMsg());
                        selectValidServer();
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                    logger.warn("Error while listening notification");
                    selectValidServer();
                }
            }
        };
    }

    /** Disable high availability mode. */
    public void disableHighAvailability() {
        enableHa = false;
        listMembersService.shutdownNow();
    }

    public EventMeta sendEvent(EventMeta event) throws Exception {
        boolean enableIdempotence =
                this.conf.getBoolean(
                        CLIENT_ENABLE_IDEMPOTENCE_CONFIG_KEY,
                        CLIENT_ENABLE_IDEMPOTENCE_CONFIG_DEFAULT_VALUE);
        String signature = UUID.randomUUID().toString();
        if (enableIdempotence) {
            int currentSeqNum = this.sequenceNum.get();
            signature =
                    StringUtils.join(Arrays.asList("client", this.clientId, currentSeqNum), "_");
        }
        NotificationServiceOuterClass.EventProto.Builder eventBuilder =
                NotificationServiceOuterClass.EventProto.newBuilder();
        if (!StringUtils.isEmpty(event.getKey())) {
            eventBuilder.setKey(event.getKey());
        }
        if (!StringUtils.isEmpty(namespace)) {
            eventBuilder.setNamespace(namespace);
        }
        if (!StringUtils.isEmpty(sender)) {
            eventBuilder.setSender(sender);
        }
        if (!StringUtils.isEmpty(event.getValue())) {
            eventBuilder.setValue(event.getValue());
        }
        if (!StringUtils.isEmpty(event.getContext())) {
            eventBuilder.setContext(event.getContext());
        }
        NotificationServiceOuterClass.SendEventRequest request =
                NotificationServiceOuterClass.SendEventRequest.newBuilder()
                        .setEvent(eventBuilder.build())
                        .setUuid(signature)
                        .setEnableIdempotence(enableIdempotence)
                        .build();
        NotificationServiceOuterClass.SendEventsResponse response =
                notificationServiceStub.sendEvent(request);
        if (response.getReturnCode() == NotificationServiceOuterClass.ReturnStatus.SUCCESS) {
            if (enableIdempotence) {
                this.sequenceNum.getAndIncrement();
            }
            return EventMeta.buildEventMeta(response.getEvent());
        } else {
            throw new Exception(response.getReturnMsg());
        }
    }

    /**
     * List specific registered listener events in Notification Service.
     *
     * @param key Key of events to be listed, null indicates not filter by name.
     * @param namespace Namespace of events to be listed, null indicates not filter by namespace.
     * @param sender Sender of events to be listed, null indicates not filter by sender.
     * @param beginOffset Offset of the events must be greater than this offset.
     * @param endOffset Offset of the events must be less than or equal to this offset.
     * @return List of event updated in Notification Service.
     */
    public List<EventMeta> listEvents(String key, String namespace, String sender, Long beginOffset, Long endOffset)
            throws Exception {

        return listEvents(
                notificationServiceStub,
                key,
                namespace,
                sender,
                beginOffset,
                endOffset,
                0);
    }
    protected static List<EventMeta> listEvents(
            NotificationServiceGrpc.NotificationServiceBlockingStub serviceStub,
            String key,
            String namespace,
            String sender,
            Long beginOffset,
            Long endOffset,
            Integer timeoutSeconds)
            throws Exception {
        NotificationServiceOuterClass.ListEventsRequest.Builder builder =
                NotificationServiceOuterClass.ListEventsRequest.newBuilder();
        if (!StringUtils.isEmpty(key)) {
            builder.setKey(key);
        }
        if (!StringUtils.isEmpty(namespace)) {
            builder.setNamespace(namespace);
        }
        if (!StringUtils.isEmpty(sender)) {
            builder.setSender(sender);
        }
        if (null != beginOffset) {
            builder.setStartOffset(beginOffset.longValue());
        }
        if (null != endOffset) {
            builder.setEndOffset(endOffset.longValue());
        }
        if (null != timeoutSeconds) {
            builder.setTimeoutSeconds(timeoutSeconds.intValue());
        }
        NotificationServiceOuterClass.ListEventsRequest request = builder.build();
        return parseEventsFromResponse(serviceStub.listEvents(request));
    }

    @Override
    public Long timeToOffset(Long timestamp) throws Exception {
        NotificationServiceOuterClass.TimeToOffsetRequest request =
                NotificationServiceOuterClass.TimeToOffsetRequest.newBuilder()
                        .setTimestamp(timestamp)
                        .build();
        NotificationServiceOuterClass.TimeToOffsetResponse response =
                notificationServiceStub.timestampToEventOffset(request);
        if (response.getReturnCode() != NotificationServiceOuterClass.ReturnStatus.SUCCESS) {
            throw new Exception(
                    "There is no event whose create_time is greater than or equal to " + timestamp);
        } else {
            return response.getOffset();
        }
    }

    /**
     * Count events in Notification Service.
     *
     * @param key Key of events to be counted, null indicates not filter by name.
     * @param namespace Namespace of events to be counted, null indicates not filter by namespace.
     * @param sender Sender of events to be counted, null indicates not filter by sender.
     * @param beginOffset Offset of the events must be greater than this offset.
     * @param endOffset Offset of the events must be less than or equal to this offset.
     * @return Count of events in Notification Service.
     */
    public ImmutablePair<Long, List<SenderEventCount>> countEvents(
            String key, String namespace, String sender, Long beginOffset, Long endOffset)
            throws Exception {

        NotificationServiceOuterClass.CountEventsRequest.Builder builder =
                NotificationServiceOuterClass.CountEventsRequest.newBuilder();
        if (!StringUtils.isEmpty(key)) {
            builder.setKey(key);
        }
        if (!StringUtils.isEmpty(namespace)) {
            builder.setNamespace(namespace);
        }
        if (!StringUtils.isEmpty(sender)) {
            builder.setSender(sender);
        }
        if (null != beginOffset) {
            builder.setStartOffset(beginOffset.longValue());
        }
        if (null != endOffset) {
            builder.setEndOffset(endOffset.longValue());
        }
        NotificationServiceOuterClass.CountEventsRequest request = builder.build();
        return parseEventCountFromResponse(notificationServiceStub.countEvents(request));
    }

    /**
     * List all registered listener events in Notification Service.
     *
     * @param startTime (Optional) The event create time after the given startTime.
     * @param startOffset (Optional) Start offset of event for listing.
     * @param endOffset (Optional) End offset of event for listing.
     * @return List of event updated in Notification Service.
     */
    public List<EventMeta> listAllEvents(Long startTime, Long startOffset, Long endOffset)
            throws Exception {
        return listAllEvents(notificationServiceStub, startTime, startOffset, endOffset, 0);
    }

    public ListenerRegistrationId registerListener(
            ListenerProcessor listenerProcessor, List<String> eventKeys, Long offset) {
        String listenKey = String.format("%s_%s_%s", eventKeys, offset, listenerProcessor.toString());
        EventListener listener =
                new EventListener(
                        notificationServiceStub, this.namespace, eventKeys, offset, listenerProcessor, 10);
        listener.start();
        threads.put(listenKey, listener);
        return new ListenerRegistrationId(listenKey);
    }

    public void unRegisterListener(ListenerRegistrationId id) {
        String listenKey = id.getId();
        if (threads.containsKey(listenKey)) {
            threads.get(listenKey).shutdown();
            threads.remove(listenKey);
        }
    }

    private String getStringValue(String original) {
        return StringUtils.isEmpty(original) ? "" : original;
    }

    private long getLongValue(Long original) {
        return original == null ? 0 : original;
    }

    /**
     * Get latest offset of specific `key` notifications in Notification Service.
     *
     * @param namespace Namespace of the event.
     * @param key Key of the event.
     */
    public long getLatestOffset(String namespace, String key) throws Exception {
        if (StringUtils.isEmpty(key)) {
            throw new Exception("Empty event key, please provide valid event key");
        } else {
            NotificationServiceOuterClass.GetLatestOffsetByKeyRequest.Builder builder =
                    NotificationServiceOuterClass.GetLatestOffsetByKeyRequest.newBuilder();
            if (!StringUtils.isEmpty(namespace)) {
                builder.setNamespace(namespace);
            }
            NotificationServiceOuterClass.GetLatestOffsetByKeyRequest request =
                    builder.setName(key).build();
            NotificationServiceOuterClass.GetLatestOffsetResponse response =
                    notificationServiceStub.getLatestOffsetByKey(request);
            return parseLatestOffsetFromResponse(response);
        }
    }

    public long parseLatestOffsetFromResponse(
            NotificationServiceOuterClass.GetLatestOffsetResponse response) throws Exception {
        if (response.getReturnCode()
                .equals(NotificationServiceOuterClass.ReturnStatus.ERROR.toString())) {
            throw new Exception(response.getReturnMsg());
        } else {
            return response.getOffset();
        }
    }

    public AtomicInteger getSequenceNum() {
        return this.sequenceNum;
    }

    public Long getClientId() {
        return clientId;
    }

    public void close() throws Exception {
        if (this.clientId >= 0) {
            NotificationServiceOuterClass.ClientIdRequest request =
                    NotificationServiceOuterClass.ClientIdRequest.newBuilder()
                            .setClientId(this.clientId)
                            .build();
            NotificationServiceOuterClass.CommonResponse response =
                    notificationServiceStub.deleteClient(request);
            if (response.getReturnCode() != NotificationServiceOuterClass.ReturnStatus.SUCCESS) {
                throw new Exception(response.getReturnMsg());
            }
        }
        disableHighAvailability();
        channel.shutdown();
        try {
            channel.awaitTermination(5000, TimeUnit.MILLISECONDS);
        } catch (InterruptedException ex) {
            logger.error("gRPC channel shutdown interrupted");
        }
    }
}

File: lib/notification_service/notification_service/client/embedded_notification_client.py
# Copyright 2022 The AI Flow Authors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
import logging
import os
import threading
import time
import grpc
from functools import wraps
from random import shuffle
from datetime import datetime
from typing import List, Dict, Tuple


from notification_service.model.event import Event
from notification_service.client.notification_client import NotificationClient, ListenerRegistrationId, \
    ListenerProcessor
from notification_service.rpc.protobuf import notification_service_pb2_grpc
from notification_service.rpc.protobuf.notification_service_pb2 import SendEventRequest, \
    ListEventsRequest, CountEventsRequest, EventProto, ReturnStatus, ListAllEventsRequest, \
    RegisterClientRequest, ClientMeta, ClientIdRequest, TimeToOffsetRequest, ListMembersRequest
from notification_service.util.utils import event_proto_to_event, proto_to_member, sleep_and_detecting_running, \
    event_count_proto_to_event_count

from notification_service.model.sender_event_count import SenderEventCount

NOTIFICATION_TIMEOUT_SECONDS = os.environ.get("NOTIFICATION_TIMEOUT_SECONDS", 5)


class SequenceNumberManager(object):

    def __init__(self, seq_num: int = 0):
        self._lock = threading.RLock()
        self._seq_num = seq_num

    def increment_sequence_number(self):
        with self._lock:
            self._seq_num += 1

    def get_sequence_number(self):
        return self._seq_num


class EmbeddedNotificationClient(NotificationClient):
    def __init__(self,
                 server_uri: str,
                 namespace: str = None,
                 sender: str = None,
                 client_id: int = None,
                 initial_seq_num: int = None
                 ):
        super().__init__(namespace, sender)
        self.server_uri = server_uri
        channel = grpc.insecure_channel(target=self.server_uri,
                                        options=[('grpc.max_receive_message_length', -1)])
        self.notification_stub = notification_service_pb2_grpc.NotificationServiceStub(channel)
        self.threads = {}  # type: Dict[int, threading.Thread]
        self.lock = threading.Lock()
        self._client_id = client_id
        self._initial_seq_num = initial_seq_num

        self.list_member_interval_ms = 1000
        self.retry_interval_ms = 1000
        self.retry_timeout_ms = 10000

        server_uri_list = self.server_uri.split(",")
        if len(server_uri_list) > 1:
            self.living_members = []
            self.current_uri = None
            last_error = None
            for server_uri in server_uri_list:
                channel = grpc.insecure_channel(server_uri)
                notification_stub = notification_service_pb2_grpc.NotificationServiceStub(channel)
                try:
                    request = ListMembersRequest(timeout_seconds=0)
                    response = notification_stub.listMembers(request)
                    if response.return_code == ReturnStatus.SUCCESS:
                        self.living_members = [proto_to_member(proto).server_uri
                                               for proto in response.members]
                    else:
                        raise Exception(response.return_msg)
                    self.current_uri = server_uri
                    self.notification_stub = notification_stub
                    break
                except grpc.RpcError as e:
                    last_error = e
            if self.current_uri is None:
                raise Exception("No available server uri!") from last_error
            self.ha_change_lock = threading.Lock()
            self.ha_running = True
            self.notification_stub = self._wrap_rpcs(self.notification_stub, server_uri)
            self.list_member_thread = threading.Thread(target=self._list_members, daemon=True)
            self.list_member_thread.start()

        if self._client_id is None:
            request = RegisterClientRequest(
                client_meta=ClientMeta(namespace=self.namespace, sender=self.sender))
            response = self.notification_stub.registerClient(request)
            if response.return_code == ReturnStatus.SUCCESS:
                self._client_id = response.client_id
            else:
                raise Exception(response.return_msg)
        else:
            request = ClientIdRequest(client_id=self._client_id)
            response = self.notification_stub.isClientExists(request)
            if response.return_code != ReturnStatus.SUCCESS:
                raise Exception("Failed to close notification client: {}".format(self))
            elif not response.is_exists:
                raise Exception("Init notification client with a client id which have not registered.")
        seq_num = 0 if self._initial_seq_num is None else int(self._initial_seq_num)
        self.sequence_num_manager = SequenceNumberManager(seq_num)

    def _ha_wrapper(self, func):
        @wraps(func)
        def call_with_retry(*args, **kwargs):
            current_func = getattr(self.notification_stub,
                                   func.__name__).inner_func
            start_time = time.time_ns() / 1000000
            failed_members = set()
            while True:
                try:
                    return current_func(*args, **kwargs)
                except grpc.RpcError:
                    logging.error("Exception thrown when calling rpc, change the connection.",
                                  exc_info=True)
                    with self.ha_change_lock:
                        # check the current_uri to ensure thread safety
                        if current_func.server_uri == self.current_uri:
                            living_members = list(self.living_members)
                            failed_members.add(self.current_uri)
                            shuffle(living_members)
                            found_new_member = False
                            for server_uri in living_members:
                                if server_uri in failed_members:
                                    continue
                                next_uri = server_uri
                                channel = grpc.insecure_channel(next_uri)
                                notification_stub = self._wrap_rpcs(
                                    notification_service_pb2_grpc.NotificationServiceStub(channel),
                                    next_uri)
                                self.notification_stub = notification_stub
                                current_func = getattr(self.notification_stub,
                                                       current_func.__name__).inner_func
                                self.current_uri = next_uri
                                found_new_member = True
                            if not found_new_member:
                                logging.error("No available living members currently. Sleep and retry.")
                                failed_members.clear()
                                sleep_and_detecting_running(self.retry_interval_ms,
                                                            lambda: self.ha_running)

                # break if stopped or timeout
                if not self.ha_running or \
                        time.time_ns() / 1000000 > start_time + self.retry_timeout_ms:
                    if not self.ha_running:
                        raise Exception("HA has been disabled.")
                    else:
                        raise Exception("Rpc retry timeout!")

        call_with_retry.inner_func = func
        return call_with_retry

    def _wrap_rpcs(self, stub, server_uri):
        for method_name, method in dict(stub.__dict__).items():
            method.__name__ = method_name
            method.server_uri = server_uri
            setattr(stub, method_name, self._ha_wrapper(method))
        return stub

    def _list_members(self):
        while self.ha_running:
            # refresh the living members
            request = ListMembersRequest(timeout_seconds=int(self.list_member_interval_ms / 1000))
            response = self.notification_stub.listMembers(request)
            if response.return_code == ReturnStatus.SUCCESS:
                with self.ha_change_lock:
                    self.living_members = [proto_to_member(proto).server_uri
                                           for proto in response.members]
            else:
                logging.error("Exception thrown when updating the living members: %s" %
                              response.return_msg)
            time.sleep(int(self.list_member_interval_ms / 1000))

    def disable_high_availability(self):
        if hasattr(self, "ha_running"):
            self.ha_running = False
            self.list_member_thread.join()

    def close(self):
        if self._client_id is not None:
            request = ClientIdRequest(client_id=self._client_id)
            response = self.notification_stub.deleteClient(request)
            if response.return_code != ReturnStatus.SUCCESS:
                raise Exception("Failed to close notification client: {}".format(self))
        logging.info("The notification client:{} has been closed.".format(self))

    @property
    def client_id(self):
        """Int. Id of the Client."""
        return self._client_id

    @property
    def sequence_number(self):
        return self.sequence_num_manager.get_sequence_number()

    def send_event(self, event: Event):
        """
        Send event to Notification Service.

        :param event: the event updated.
        :return: The created event which has offset and create time.
        """
        seq_num = self.sequence_num_manager.get_sequence_number()
        signature = '_'.join(['client', str(self._client_id), str(seq_num)])

        request = SendEventRequest(
            event=EventProto(
                key=event.key,
                value=event.value,
                context=event.context,
                namespace=self.namespace,
                sender=self.sender
            ),
            uuid=signature,
            enable_idempotence=True)
        response = self.notification_stub.sendEvent(request)
        if response.return_code == ReturnStatus.SUCCESS:
            self.sequence_num_manager.increment_sequence_number()
            return event_proto_to_event(response.event)
        else:
            raise Exception(response.return_msg)

    def register_listener(self,
                          listener_processor: ListenerProcessor,
                          event_keys: List[str] = None,
                          offset: int = None) -> ListenerRegistrationId:

        def list_events_from_offset(client, v, timeout_seconds: int = None):
            request = ListAllEventsRequest(start_offset=v, timeout_seconds=timeout_seconds)
            response = client.notification_stub.listAllEvents(request)
            if response.return_code == ReturnStatus.SUCCESS:
                if response.events is None:
                    return None
                else:
                    events = []
                    for event_proto in response.events:
                        event = event_proto_to_event(event_proto)
                        events.append(event)
                    return events
            else:
                raise Exception(response.return_msg)

        def listen(client, v, p):
            t = threading.current_thread()
            current_offset = 0 if v is None else v
            try:
                while getattr(t, '_flag', True):
                    notifications = list_events_from_offset(client,
                                                            current_offset,
                                                            NOTIFICATION_TIMEOUT_SECONDS)
                    if len(notifications) > 0:
                        if event_keys is not None:
                            events = client.filter_events(event_keys=event_keys, events=notifications)
                        else:
                            events = notifications
                        p.process(events)
                        current_offset = notifications[len(notifications) - 1].offset
            except Exception as e:
                logging.exception("Exception when listening events, %s", e)
                raise e

        thread = threading.Thread(target=listen,
                                  args=(self, offset, listener_processor),
                                  daemon=True)
        thread.start()
        self.lock.acquire()
        try:
            self.threads[thread.__hash__()] = thread
        finally:
            self.lock.release()
        return ListenerRegistrationId(id=str(thread.__hash__()))

    def unregister_listener(self, registration_id: ListenerRegistrationId):
        self.lock.acquire()
        key = int(registration_id.id)
        try:
            if key in self.threads:
                thread = self.threads[key]
                thread._flag = False
                thread.join()
                del self.threads[key]
        finally:
            self.lock.release()

    def list_all_events(self,
                        start_time: int = None,
                        start_offset: int = None,
                        end_offset: int = None) -> List[Event]:
        """
        List the range of `offset` of events in Notification Service.

        :param start_time: (Optional) Start time of the events.
        :param start_offset: (Optional) the offset of the events must greater than the
                              start_offset.
        :param end_offset: (Optional) the offset of the events must equal or less than the
                            end_offset.
        :return: The event list.
        """
        request = ListAllEventsRequest(start_time=start_time,
                                       start_offset=start_offset,
                                       end_offset=end_offset)
        response = self.notification_stub.listAllEvents(request)
        if response.return_code == ReturnStatus.SUCCESS:
            if response.events is None:
                return []
            else:
                events = []
                for event_proto in response.events:
                    event = event_proto_to_event(event_proto)
                    events.append(event)
                return events
        else:
            raise Exception(response.return_msg)

    def list_events(self,
                    key: str = None,
                    namespace: str = None,
                    sender: str = None,
                    begin_offset: int = None,
                    end_offset: int = None) -> List[Event]:
        """
        List specific events in Notification Service.

        :param key: Key of the event for listening.
        :param namespace: Namespace of the event for listening.
        :param sender: The event sender.
        :param begin_offset: Offset of the events must be greater than this offset.
        :param end_offset: Offset of the events must be less than or equal to this offset.
        :return: The event list.
        """
        request = ListEventsRequest(
            key=key,
            namespace=namespace,
            sender=sender,
            start_offset=begin_offset,
            end_offset=end_offset,
        )
        response = self.notification_stub.listEvents(request)
        if response.return_code == ReturnStatus.SUCCESS:
            if response.events is None:
                return []
            else:
                events = []
                for event_proto in response.events:
                    event = event_proto_to_event(event_proto)
                    events.append(event)
                return events
        else:
            raise Exception(response.return_msg)

    def time_to_offset(self, time: datetime) -> int:
        timestamp = int(time.timestamp() * 1000)
        request = TimeToOffsetRequest(timestamp=timestamp)
        response = self.notification_stub.timestampToEventOffset(request)
        return response.offset

    def count_events(self,
                     key: str = None,
                     namespace: str = None,
                     sender: str = None,
                     begin_offset: int = None,
                     end_offset: int = None) -> Tuple[int, List[SenderEventCount]]:
        """
        Count specific events in Notification Service.

        :param key: Key of the event for listening.
        :param namespace: Namespace of the event for listening.
        :param sender: The event sender.
        :param begin_offset: Offset of the events must be greater than this offset.
        :param end_offset: Offset of the events must be less than or equal to this offset.
        :return: The total event count and the list of event counts of each sender.
        """
        request = CountEventsRequest(
            key=key,
            namespace=namespace,
            sender=sender,
            start_offset=begin_offset,
            end_offset=end_offset,
        )
        response = self.notification_stub.countEvents(request)
        if response.return_code == ReturnStatus.SUCCESS:
            sender_event_counts = []
            for sender_event_count_proto in response.sender_event_counts:
                sender_event_count = event_count_proto_to_event_count(sender_event_count_proto)
                sender_event_counts.append(sender_event_count)
            return response.event_count, sender_event_counts
        else:
            raise Exception(response.return_msg)

    def filter_events(self, event_keys: List[str], events: List[Event]) -> List[Event]:
        def match(key, namespace, event: Event) -> bool:
            if key is not None and key != event.key:
                return False
            if namespace is not None and namespace != event.namespace:
                return False
            return True

        results = []
        for e in events:
            for k in event_keys:
                if match(k, self.namespace, e):
                    results.append(e)
                    break
        return results
